File: typescript/src/session.ts
export interface RefreshCredentialFacts {
  readonly nowMs: number;
  readonly expiresAtMs: number;
  readonly rotatedAtMs?: number | null;
  readonly familyRevokedAtMs?: number | null;
  readonly gracePeriodMs: number;
}

export type RefreshCredentialDecision =
  | { readonly kind: 'rotate' }
  | { readonly kind: 'retry-current' }
  | { readonly kind: 'reject'; readonly reason: 'expired' }
  | {
      readonly kind: 'revoke-family';
      readonly reason: 'family-revoked' | 'credential-reuse';
    };

export type SessionBindingDecision = 'unbound' | 'not-checked' | 'match' | 'mismatch';

export interface SessionLifetimePolicy {
  readonly defaultSpanMs: number;
  readonly rememberedSpanMs: number;
  readonly maxIdleMs: number;
}

function requireFinite(value: number, label: string): void {
  if (!Number.isFinite(value)) throw new RangeError(`${label} must be finite`);
}

function requireNonNegative(value: number, label: string): void {
  requireFinite(value, label);
  if (value < 0) throw new RangeError(`${label} must not be negative`);
}

export function classifyRefreshCredential(
  facts: RefreshCredentialFacts,
): RefreshCredentialDecision {
  requireFinite(facts.nowMs, 'nowMs');
  requireFinite(facts.expiresAtMs, 'expiresAtMs');
  requireNonNegative(facts.gracePeriodMs, 'gracePeriodMs');
  if (facts.familyRevokedAtMs != null) {
    requireFinite(facts.familyRevokedAtMs, 'familyRevokedAtMs');
    return { kind: 'revoke-family', reason: 'family-revoked' };
  }
  if (facts.expiresAtMs <= facts.nowMs) return { kind: 'reject', reason: 'expired' };
  if (facts.rotatedAtMs == null) return { kind: 'rotate' };
  requireFinite(facts.rotatedAtMs, 'rotatedAtMs');
  if (facts.nowMs - facts.rotatedAtMs < facts.gracePeriodMs) {
    return { kind: 'retry-current' };
  }
  return { kind: 'revoke-family', reason: 'credential-reuse' };
}

export function credentialEpochMatches(observedEpoch: number, currentEpoch: number): boolean {
  return (
    Number.isSafeInteger(observedEpoch) &&
    observedEpoch >= 0 &&
    Number.isSafeInteger(currentEpoch) &&
    currentEpoch >= 0 &&
    observedEpoch === currentEpoch
  );
}

export function isRefreshRetryEligible(
  facts: RefreshCredentialFacts & {
    readonly observedCredentialEpoch: number;
    readonly currentCredentialEpoch: number;
  },
): boolean {
  return (
    credentialEpochMatches(facts.observedCredentialEpoch, facts.currentCredentialEpoch) &&
    classifyRefreshCredential(facts).kind === 'retry-current'
  );
}

export function isActiveRefreshCredential(facts: RefreshCredentialFacts): boolean {
  return classifyRefreshCredential(facts).kind === 'rotate';
}

export function evaluateSessionBinding(
  expectedBinding: string | null | undefined,
  presentedBinding: string | null | undefined,
): SessionBindingDecision {
  if (expectedBinding == null || expectedBinding === '') return 'unbound';
  if (presentedBinding == null) return 'not-checked';
  return expectedBinding === presentedBinding ? 'match' : 'mismatch';
}

export function sessionSpanMs(
  remembered: boolean | undefined,
  policy: Pick<SessionLifetimePolicy, 'defaultSpanMs' | 'rememberedSpanMs'>,
): number {
  requireNonNegative(policy.defaultSpanMs, 'defaultSpanMs');
  requireNonNegative(policy.rememberedSpanMs, 'rememberedSpanMs');
  return remembered === true ? policy.rememberedSpanMs : policy.defaultSpanMs;
}

export function deriveSessionSpanMs(createdAtMs: number, expiresAtMs: number): number {
  requireFinite(createdAtMs, 'createdAtMs');
  requireFinite(expiresAtMs, 'expiresAtMs');
  return expiresAtMs - createdAtMs;
}

export function sessionIdleWindowMs(spanMs: number, maxIdleMs: number): number {
  requireNonNegative(spanMs, 'spanMs');
  requireNonNegative(maxIdleMs, 'maxIdleMs');
  return Math.min(spanMs, maxIdleMs);
}

export function isSessionActive(revokedAtMs: number | null): boolean {
  return revokedAtMs === null;
}

export function isSessionRevoked(revokedAtMs: number | null): boolean {
  return !isSessionActive(revokedAtMs);
}

export function sessionAgeMs(createdAtMs: number, nowMs: number): number {
  requireFinite(createdAtMs, 'createdAtMs');
  requireFinite(nowMs, 'nowMs');
  return nowMs - createdAtMs;
}

export function isSessionIdle(
  lastActiveAtMs: number,
  idleTimeoutMs: number,
  nowMs: number,
): boolean {
  requireFinite(lastActiveAtMs, 'lastActiveAtMs');
  requireNonNegative(idleTimeoutMs, 'idleTimeoutMs');
  requireFinite(nowMs, 'nowMs');
  return nowMs - lastActiveAtMs > idleTimeoutMs;
}

export function sessionIdleRemainingMs(
  lastActiveAtMs: number,
  idleTimeoutMs: number,
  nowMs: number,
): number {
  requireFinite(lastActiveAtMs, 'lastActiveAtMs');
  requireNonNegative(idleTimeoutMs, 'idleTimeoutMs');
  requireFinite(nowMs, 'nowMs');
  return Math.max(0, idleTimeoutMs - (nowMs - lastActiveAtMs));
}

export function selectSessionsForEviction<Session>(
  sessions: readonly Session[],
  maxSessions: number,
  createdAtMs: (session: Session) => number,
): readonly Session[] {
  if (!Number.isSafeInteger(maxSessions)) {
    throw new RangeError('maxSessions must be a safe integer');
  }
  if (maxSessions <= 0 || sessions.length <= maxSessions) return [];
  const sorted = [...sessions].sort((left, right) => {
    const leftCreatedAt = createdAtMs(left);
    const rightCreatedAt = createdAtMs(right);
    requireFinite(leftCreatedAt, 'session createdAtMs');
    requireFinite(rightCreatedAt, 'session createdAtMs');
    return leftCreatedAt - rightCreatedAt;
  });
  return sorted.slice(0, sessions.length - maxSessions);
}

File: rust/src/session.rs
use crate::{AuthError, AuthResult};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct RefreshCredentialFacts {
    pub now_ms: i64,
    pub expires_at_ms: i64,
    pub rotated_at_ms: Option<i64>,
    pub family_revoked_at_ms: Option<i64>,
    pub grace_period_ms: u64,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum RefreshCredentialDecision {
    Rotate,
    RetryCurrent,
    RejectExpired,
    RevokeFamilyRevoked,
    RevokeCredentialReuse,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum SessionBindingDecision {
    Unbound,
    NotChecked,
    Match,
    Mismatch,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct SessionLifetimePolicy {
    pub default_span_ms: u64,
    pub remembered_span_ms: u64,
    pub max_idle_ms: u64,
}

#[must_use]
pub fn classify_refresh_credential(facts: RefreshCredentialFacts) -> RefreshCredentialDecision {
    if facts.family_revoked_at_ms.is_some() {
        return RefreshCredentialDecision::RevokeFamilyRevoked;
    }
    if facts.expires_at_ms <= facts.now_ms {
        return RefreshCredentialDecision::RejectExpired;
    }
    let Some(rotated_at_ms) = facts.rotated_at_ms else {
        return RefreshCredentialDecision::Rotate;
    };
    let elapsed = i128::from(facts.now_ms) - i128::from(rotated_at_ms);
    if elapsed < i128::from(facts.grace_period_ms) {
        RefreshCredentialDecision::RetryCurrent
    } else {
        RefreshCredentialDecision::RevokeCredentialReuse
    }
}

#[must_use]
pub const fn credential_epoch_matches(observed_epoch: u64, current_epoch: u64) -> bool {
    observed_epoch == current_epoch
}

#[must_use]
pub fn is_refresh_retry_eligible(
    facts: RefreshCredentialFacts,
    observed_credential_epoch: u64,
    current_credential_epoch: u64,
) -> bool {
    credential_epoch_matches(observed_credential_epoch, current_credential_epoch)
        && classify_refresh_credential(facts) == RefreshCredentialDecision::RetryCurrent
}

#[must_use]
pub fn is_active_refresh_credential(facts: RefreshCredentialFacts) -> bool {
    classify_refresh_credential(facts) == RefreshCredentialDecision::Rotate
}

#[must_use]
pub fn evaluate_session_binding(
    expected_binding: Option<&str>,
    presented_binding: Option<&str>,
) -> SessionBindingDecision {
    match (expected_binding, presented_binding) {
        (None | Some(""), _) => SessionBindingDecision::Unbound,
        (Some(_), None) => SessionBindingDecision::NotChecked,
        (Some(expected), Some(presented)) if expected == presented => SessionBindingDecision::Match,
        (Some(_), Some(_)) => SessionBindingDecision::Mismatch,
    }
}

#[must_use]
pub const fn session_span_ms(remembered: Option<bool>, policy: SessionLifetimePolicy) -> u64 {
    if matches!(remembered, Some(true)) {
        policy.remembered_span_ms
    } else {
        policy.default_span_ms
    }
}

/// Derives the issued session span from creation and expiration timestamps.
///
/// # Errors
///
/// Returns [`AuthError::ArithmeticOverflow`] when the difference exceeds `i64`.
pub fn derive_session_span_ms(created_at_ms: i64, expires_at_ms: i64) -> AuthResult<i64> {
    expires_at_ms
        .checked_sub(created_at_ms)
        .ok_or(AuthError::ArithmeticOverflow("session span overflowed i64"))
}

#[must_use]
pub const fn session_idle_window_ms(span_ms: u64, max_idle_ms: u64) -> u64 {
    if span_ms < max_idle_ms {
        span_ms
    } else {
        max_idle_ms
    }
}

#[must_use]
pub const fn is_session_active(revoked_at_ms: Option<i64>) -> bool {
    revoked_at_ms.is_none()
}

#[must_use]
pub const fn is_session_revoked(revoked_at_ms: Option<i64>) -> bool {
    !is_session_active(revoked_at_ms)
}

/// Calculates session age at an injected reference time.
///
/// # Errors
///
/// Returns [`AuthError::ArithmeticOverflow`] when the difference exceeds `i64`.
pub fn session_age_ms(created_at_ms: i64, now_ms: i64) -> AuthResult<i64> {
    now_ms
        .checked_sub(created_at_ms)
        .ok_or(AuthError::ArithmeticOverflow("session age overflowed i64"))
}

#[must_use]
pub fn is_session_idle(last_active_at_ms: i64, idle_timeout_ms: u64, now_ms: i64) -> bool {
    i128::from(now_ms) - i128::from(last_active_at_ms) > i128::from(idle_timeout_ms)
}

#[must_use]
pub fn session_idle_remaining_ms(last_active_at_ms: i64, idle_timeout_ms: u64, now_ms: i64) -> u64 {
    let remaining =
        i128::from(idle_timeout_ms) - (i128::from(now_ms) - i128::from(last_active_at_ms));
    if remaining <= 0 {
        0
    } else {
        u64::try_from(remaining).unwrap_or(u64::MAX)
    }
}

#[must_use]
pub fn select_sessions_for_eviction<T, F>(
    sessions: &[T],
    max_sessions: usize,
    created_at_ms: F,
) -> Vec<&T>
where
    F: Fn(&T) -> i64,
{
    if max_sessions == 0 || sessions.len() <= max_sessions {
        return Vec::new();
    }
    let mut ordered = sessions.iter().collect::<Vec<_>>();
    ordered.sort_by_key(|session| created_at_ms(session));
    ordered.truncate(sessions.len() - max_sessions);
    ordered
}
